File: bindings/lua/proven/safe_math.lua
-- SPDX-License-Identifier: PMPL-1.0
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe arithmetic operations that cannot crash or overflow unexpectedly.
--- @module proven.safe_math

local safe_math = {}

-- Lua 5.3+ integer limits (simulated for bounded operations)
local MAX_INT = 9223372036854775807
local MIN_INT = -9223372036854775808

--- Safely divide two integers, returning nil on division by zero.
--- @param numerator number
--- @param denominator number
--- @return number|nil
function safe_math.div(numerator, denominator)
    if denominator == 0 then
        return nil
    end
    return math.floor(numerator / denominator)
end

--- Safely compute modulo, returning nil on division by zero.
--- @param numerator number
--- @param denominator number
--- @return number|nil
function safe_math.mod(numerator, denominator)
    if denominator == 0 then
        return nil
    end
    return numerator % denominator
end

--- Safely add two integers, returning nil on overflow.
--- @param a number
--- @param b number
--- @return number|nil
function safe_math.add(a, b)
    local result = a + b
    if result > MAX_INT or result < MIN_INT then
        return nil
    end
    return result
end

--- Safely subtract two integers, returning nil on overflow.
--- @param a number
--- @param b number
--- @return number|nil
function safe_math.sub(a, b)
    local result = a - b
    if result > MAX_INT or result < MIN_INT then
        return nil
    end
    return result
end

--- Safely multiply two integers, returning nil on overflow.
--- @param a number
--- @param b number
--- @return number|nil
function safe_math.mul(a, b)
    local result = a * b
    if result > MAX_INT or result < MIN_INT then
        return nil
    end
    return result
end

--- Checked add that raises error on overflow.
--- @param a number
--- @param b number
--- @return number
function safe_math.checked_add(a, b)
    local result = safe_math.add(a, b)
    if result == nil then
        error("integer overflow")
    end
    return result
end

--- Checked subtract that raises error on overflow.
--- @param a number
--- @param b number
--- @return number
function safe_math.checked_sub(a, b)
    local result = safe_math.sub(a, b)
    if result == nil then
        error("integer overflow")
    end
    return result
end

--- Checked multiply that raises error on overflow.
--- @param a number
--- @param b number
--- @return number
function safe_math.checked_mul(a, b)
    local result = safe_math.mul(a, b)
    if result == nil then
        error("integer overflow")
    end
    return result
end

return safe_math

File: bindings/elixir/lib/proven/safe_math.ex
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

defmodule Proven.SafeMath do
  @moduledoc """
  Safe arithmetic operations via libproven FFI.

  All computation is performed in Idris 2 verified code.
  These functions cannot crash or overflow unexpectedly.
  """

  @doc """
  Safely divide two integers.

  Returns `{:ok, result}` or `{:error, :division_by_zero}`.
  """
  @spec safe_div(integer(), integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_div(numerator, denominator) when is_integer(numerator) and is_integer(denominator) do
    case Proven.NIF.nif_math_div(numerator, denominator) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :division_by_zero}
    end
  end

  @doc """
  Safely compute modulo.

  Returns `{:ok, result}` or `{:error, :division_by_zero}`.
  """
  @spec safe_mod(integer(), integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_mod(numerator, denominator) when is_integer(numerator) and is_integer(denominator) do
    case Proven.NIF.nif_math_mod(numerator, denominator) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :division_by_zero}
    end
  end

  @doc """
  Safely add two integers with overflow detection.

  Returns `{:ok, result}` or `{:error, :overflow}`.
  """
  @spec safe_add(integer(), integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_add(a, b) when is_integer(a) and is_integer(b) do
    case Proven.NIF.nif_math_add(a, b) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :overflow}
    end
  end

  @doc """
  Safely subtract two integers with overflow detection.

  Returns `{:ok, result}` or `{:error, :overflow}`.
  """
  @spec safe_sub(integer(), integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_sub(a, b) when is_integer(a) and is_integer(b) do
    case Proven.NIF.nif_math_sub(a, b) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :overflow}
    end
  end

  @doc """
  Safely multiply two integers with overflow detection.

  Returns `{:ok, result}` or `{:error, :overflow}`.
  """
  @spec safe_mul(integer(), integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_mul(a, b) when is_integer(a) and is_integer(b) do
    case Proven.NIF.nif_math_mul(a, b) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :overflow}
    end
  end

  @doc """
  Safe absolute value that handles MIN_INT correctly.

  Returns `{:ok, result}` or `{:error, :overflow}`.
  """
  @spec safe_abs(integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_abs(n) when is_integer(n) do
    case Proven.NIF.nif_math_abs(n) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :overflow}
    end
  end

  @doc """
  Clamp a value to [lo, hi].
  """
  @spec clamp(integer(), integer(), integer()) :: integer()
  def clamp(lo, hi, value) when is_integer(lo) and is_integer(hi) and is_integer(value) do
    Proven.NIF.nif_math_clamp(lo, hi, value)
  end

  @doc """
  Integer power with overflow checking.

  Returns `{:ok, result}` or `{:error, :overflow}`.
  """
  @spec safe_pow(integer(), non_neg_integer()) :: {:ok, integer()} | {:error, atom()}
  def safe_pow(base, exp) when is_integer(base) and is_integer(exp) and exp >= 0 do
    case Proven.NIF.nif_math_pow(base, exp) do
      {:ok, value} -> {:ok, value}
      {:error, _status} -> {:error, :overflow}
    end
  end
end
